api/app/controllers/gift_cards_controller.rb

```diff
@@ -1,8 +1,63 @@
+require 'stripe'
+
 class GiftCardsController < ApplicationController
+  before_action :set_gift_card, only: [:show, :update, :destroy]
 
   # POST /gift_cards
   def create
-    obj = { message: 'hello world!' }
-    json_response(obj)
+    @gift_card = GiftCard.create!(create_params)
+    json_response(@gift_card, :created)
+  end
+
+  # GET /gift_cards/:d
+  def show
+    json_response(@gift_card)
+  end
+
+  # PUT /gift_cards/:id
+  def update
+    # only allows updating amounts of gift cards
+    original_amount = @gift_card.amount
+    new_amount = update_params[:amount].to_i
+
+    if original_amount < new_amount
+      json_response({ message: 'Cannot increase gift card amount' }, :unprocessable_entity)
+      return
+    end
+
+    @gift_card.update!(update_params)
+    head :no_content
+  end
+
+  def destroy
+    @gift_card.destroy
+    head :no_content
+  end
+
+  private
+
+  def create_params
+    params.require(:charge_id)
+    charge = stripe_charge
+    {
+        charge_id: params[:charge_id],
+        merchant_id: charge[:metadata][:merchant_id],
+        customer_id: charge[:customer],
+        amount: charge[:display_items].first[:amount]
+    }
+  end
+
+  def update_params
+    params.require(:amount)
+    params.permit(:amount)
+  end
+
+  def set_gift_card
+    @gift_card = GiftCard.find(params[:id])
```

Comment:
You would need to find_by(gift_card_id:params[:id])

Reply:
also you'd need to make sure that params[:id] exists / is required

Reply:
id is an integer and used internally, gift_card_id would be that long hash that we generate for each gift card upon creation